### ISSUE TITLE
CLA Assistant Status Occasionally Not Updated

### ISSUE THREAD

**Issue by winder** [cla-assistant/cla-assistant#520]
Today we noticed a similar issue to #512 happening again. The behavior is slightly different though, I don't see any failed webhooks like before. I do see that the hook was sent to CLAAssistant, and a 200 status was returned.

Could there be a new issue with this?

![image](https://user-images.githubusercontent.com/125509/71736331-ca879080-2e1e-11ea-9dbe-4d1fb27c8118.png)

**Comment by ibakshay** [comment]
Hi @winder, thank you for reaching to us again. Next time when this issue occurs, could you please give us  the URL of the affected Pull Request as it will help us for better debugging.

**Comment by timbussmann** [comment]
we are running into the same issue from time to time as well, here's a PR which never got a response from cla check: https://github.com/Particular/NServiceBus/pull/5549

**Comment by andreasohlund** [comment]
I can confirm that the webhook was delivered successfully in the [case mentioned by Tim](https://github.com/cla-assistant/cla-assistant/issues/520#issuecomment-571537393)

**Comment by andreasohlund** [comment]
Just tried to deliver the webhook and it succeeded again, this time is worked though and the CLA check is now green.

Are there any server-side looks we can look at to figure out why the first call didn't make the check green?

**Comment by winder** [comment]
@ibakshay sorry for the incomplete bug report, here is the PR:
https://github.com/algorand/go-algorand/pull/679

We have already worked around this particular instance by clicking the check PRs button in the CLA Assistant web interface.

**Comment by ibakshay** [comment]
Thank you for the detailed analyse of the issue. It looks like a bug from our side and we are looking into it. <br/> As a workaround, like @winder  mentioned, you can retrigger the CLA check with the `recheckPR` button from  the CLA Assistant UI, or you can also  retrigger the CLA check for a particular Pull Request using this API ` https://cla-assistant.io/check/<orgname>/<reponame>?pullRequest=10`.  @andreasohlund, In your case the recheck API will be `https://cla-assistant.io/check/ Particular/NServiceBus?pullRequest=5549`

**Comment by nycnewman** [comment]
Just FYI but we just saw a similar issues on our deployment of CLA Assistant. In this case we saw logs that suggests that the access account used by CLA Assistant is hitting a Github API rate limit (i.e. 5000 calls / hour). Until this clears, webhooks deliver correctly but we saw issues with status check updates on PRs and the CLA UI not refreshing configured repo details. 

Couple of items in logs of interest - we see a significant number of token refreshes being made for oAuth tokens when API limit clears; we see deprecation notices for a function in OctaKit; we also see two calls to API (two log entries to be precise) when a user refreshes the UI screen. So something (combination of number of repos and open PRS??) results in high GH API calls that blocks when exceeded.

**Comment by nycnewman** [comment]
Additional note on above - issue with appear to be random as exceeding rate limit cannot be predicted and time to clear may also vary depending on shortly after each hour the limit is exceeded.

**Comment by ibakshay** [comment]
@nycnewman, I am really sorry for the late response. thank you very much for taking your time in giving us the insights.  These are all very interesting insights which we will look into it, and we would really appreciate If you can contribute.  If not, also it is okay as you have helped us a lot already 😄 .

**Comment by stevenvachon** [comment]
Happening here: https://github.com/grafana/grafana/pull/22278

**Comment by bharney0** [comment]
@ibakshay We are also experiencing a similar problem on [azure-docs repo](https://github.com/MicrosoftDocs/azure-docs) on [Pull Request #44149](https://github.com/MicrosoftDocs/azure-docs/pull/44149). We have a user who has [previously signed the CLA](https://github.com/MicrosoftDocs/azure-docs/pull/39849).  But is now showing that they need to sign again. We will ask the user to do the recheck, but we are not sure what the problem is and why the CLA Assistant is hanging. Can you look into this? @ktoliver @jkuehler

**Comment by KharitonOff** [comment]
@bharney0 in your case this seems to be a different issue: the status is reported correctly, the user doesn't seem to have accepted the CLA yet. The PR https://github.com/MicrosoftDocs/azure-docs/pull/39849 passed CLA status because there were no significant changes (this is my guess according to status text `All CLA requirements met`, otherwise it would sound like `All committers have signed the CLA`).

**Comment by nycnewman** [comment]
@ibakshay any luck with your investigations. We're seeing check getting hung more frequently and appears to be API limits related. We're seeing logs with "API rate Limit exceeded for User...." daily at the moment. Is there a way to track count of API use? Is there a particular section of code worth looking at that might drive API usage up?

**Comment by KharitonOff** [comment]
@nycnewman do you run the latest version of CLA assistant? We have recently tried to reduce number of api calls by caching api call to get gist content (this is the most common call, which is done with the token of our own github user). If you would like to track the count of API use, I would suggest to do so here: https://github.com/cla-assistant/cla-assistant/blob/83cea5561f3c8c2952afa417d0c2935c6f3ed0f9/src/server/services/github.js#L27 
Most API calls (from the backend) are done here. If I find some time today or tomorrow I will try to add logging here for all failing calls with rate limit number. Maybe this will bring us closer to the root cause. 
Thank you very much for your patience and advice :)

**Comment by ktoliver** [comment]
> @bharney0 in your case this seems to be a different issue: the status is reported correctly, the user doesn't seem to have accepted the CLA yet. The PR [MicrosoftDocs/azure-docs#39849](https://github.com/MicrosoftDocs/azure-docs/pull/39849) passed CLA status because there were no significant changes (this is my guess according to status text `All CLA requirements met`, otherwise it would sound like `All committers have signed the CLA`).

Hi @KharitonOff - Does this mean that the CLA is checked only if a revision of a certain size or percentage is made in the pull request? What is the cutoff?

Also, re: the direct link to check a contributor's CLA status, when I create this link for this issue/PR, the link takes me to my GitHub dashboard. There is no message about the contributor, which is what I'd expect to see: https://cla-assistant.io/check/MicroftDocs/azure-docs?pullRequest=44149

Thanks!

**Comment by KharitonOff** [comment]
Hi @ktoliver , yes you understood it right. I don't know what the cutoff is, because this can be optionally configured for each project. 

The check link was actually designed as an API endpoint with no response. I agree with you, that it would be helpful to respond with a specific information about the status of the PR.

**Comment by winder** [comment]
I'm still seeing this issue from time to time.

https://github.com/algorand/go-algorand/pull/1687

**Comment by Halamix2** [comment]
We have issue with https://github.com/kyma-project/third-party-images/pull/204,
https://cla-assistant.io/check/kyma-project/third-party-images?pullRequest=204 doesn't help

**Comment by muthumanik** [comment]
Hi I have issue on CLA Sign in . actually i did multiple time signin CLA but still it is pending. can you please help me to resolve this. i raised pull request for rdkcentral/rdkservices but CLA is not yet signed  issues raising after signed the cla agreement multiple time.

**Comment by KharitonOff** [comment]
@muthumanik it seems to be solved now, right? At least the last PR I could find has a successful CLA check.

**Comment by Sawthis** [comment]
@KharitonOff no, it's not fixed. One example: https://github.com/kyma-project/test-infra/pull/12152#issuecomment-2415954228

You can find more examples here: https://github.com/search?q=org%3Akyma-project+%22Successfully+reached+out+to+cla-assistant.io+to+initialize+recheck%22&type=pullrequests&s=created&o=desc

**Comment by marksalpeter** [comment]
For anyone interested, calling the endpoint in #528 is a nice work around. [Here is the GH Action I added to our project](https://github.com/langfuse/langfuse/pull/13021/changes#diff-b8c117a8cd8a3f8e11c17888412b00b72db56701b7c5584d588d5699c0043757) to allow engineers to re-trigger the license check directly from a PR if needed.
